src/olmo_core/data/numpy_dataset.py

```diff
@@ -1638,3 +1608,246 @@ def build(self) -> NumpyDatasetBase:
+                futures = []
+                for path, idx in paths_needed:
+                    indices_path = self._get_indices_path(path)
+                    log.info(f"Gathering instance indices for '{path}'...")
+                    # NOTE: We limit the number of instances by total target token count // sequence length
+                    max_instances = (
+                        self._path_offset_index[(str(path), idx)] // self.sequence_length
+                    )
+                    future = executor.submit(
+                        run_worker_func,
+                        segment_documents_into_instances,
+                        path,
+                        indices_path,
+                        max_sequence_length=self.sequence_length,
+                        eos_token_id=self.eos_token_id,
+                        dtype=self.dtype,
+                        indices_dtype=self.dtype,
+                        max_instances=max_instances,
+                    )
+                    futures.append(future)
+
+                concurrent.futures.wait(futures, return_when="ALL_COMPLETED")
+
+                # Log results.
+                for path, future in zip([item[0] for item in paths_needed], futures):
+                    _, total_instances = future.result()
+                    log.info(
+                        f"Created {total_instances:,d} instances of sequence length up to "
+                        f"{self.sequence_length} from '{path}'"
+                    )
+
+    def _get_file_size_and_length(
+        self, item: Tuple[PathOrStr, int], dtype: Optional[SupportedDType] = None
+    ) -> Tuple[int, int]:
+        path, idx = item
+        dtype = dtype or self.dtype
+        item_size = dtype(0).itemsize
+        file_size = self._get_size_from_offset_index(item)
+        if (
+            self.max_target_sequence_length is None
+            or self.max_target_sequence_length == self.sequence_length
+        ):
+            return file_size, file_size // (item_size * self.sequence_length)
+        elif self.max_target_sequence_length > self.sequence_length:
+            num_max_seq_len_instances = file_size // (item_size * self.max_target_sequence_length)
+            return (
+                file_size,
+                num_max_seq_len_instances
+                * (self.max_target_sequence_length // self.sequence_length),
+            )
+        else:
+            raise RuntimeError("invalid 'max_target_sequence_length' or 'sequence_length'")
+
+    def _get_size_from_offset_index(self, path_index: Tuple[PathOrStr, int]) -> int:
+        try:
+            path, idx = path_index
+            # Get size in bytes from tokens in the supplied index * itemsize
+            return self._path_offset_index[(str(path), idx)] * self.dtype(0).itemsize
+        except KeyError:
+            raise OLMoEnvironmentError(f"Item not found in path index @ {path_index}")
+
+
+@dataclass
+class NumpyFSLDatasetMixtureConfig(Config):
```

Comment:
Is it necessary to have a separate config class, or can you just merge this into `NumpyDatasetConfig`?

Reply:
I assume this can be merged. On it.